dataplane/src/nat/mod.rs

```diff
@@ -0,0 +1,418 @@
+    fn insert_vpc(&mut self, vni: Vni, vpc: Vpc) {
+        vpc.iter_pifs().for_each(|pif| {
+            pif.iter_ips().for_each(|prefix| {
+                let _ = self.global_pif_trie.insert(prefix, pif.name().clone());
+            });
+        });
+        let _ = self.vpcs.insert(vni.as_u32(), vpc);
+    }
+
+    #[tracing::instrument(level = "trace")]
+    fn find_pif_by_ip(&self, ip: &IpAddr) -> Option<String> {
+        self.global_pif_trie.find_ip(ip)
+    }
+
+    #[tracing::instrument(level = "trace")]
+    fn find_pif_in_vpc(&self, vni: Vni, ip: &IpAddr) -> Option<String> {
+        let vpc = self.vpcs.get(&vni.as_u32())?;
+        vpc.find_pif_by_endpoint(ip)
+    }
+
+    #[tracing::instrument(level = "trace")]
+    fn get_vpc(&self, vni: Vni) -> Option<&Vpc> {
+        self.vpcs.get(&vni.as_u32())
+    }
+
+    #[tracing::instrument(level = "trace")]
+    fn find_pif_by_name(&self, name: &String) -> Option<&Pif> {
+        self.vpcs.values().find_map(|vpc| vpc.get_pif(name))
+    }
+
+    #[tracing::instrument(level = "trace")]
+    fn find_src_pif(&self, src_vpc_vni: Vni, dst_pif: &Pif, dst_ip: &IpAddr) -> Option<&Pif> {
+        // Iterate on destination PIF's peering policies
+        for peering_name in dst_pif.iter_peerings() {
+            let peering = self.peerings.get(peering_name)?;
+            let peer_pif_idx = peering.get_peer_index(dst_pif);
+            let peer_pif_vni = peering.vnis()[peer_pif_idx];
+
+            // Filter peering policies, discard if not attached to source VPC
+            if peer_pif_vni != src_vpc_vni {
+                continue;
+            }
+
+            // Retrieve destination PIF's peer PIF for the policy
+            let src_vpc = self.get_vpc(src_vpc_vni)?;
+            let peer_pif_name = &peering.pifs()[peer_pif_idx];
+            let peer_pif = src_vpc.get_pif(peer_pif_name)?;
+
+            // Search peer PIF's endpoints for packet's destination IP
+            if peer_pif
+                .iter_endpoints()
+                .any(|endpoint| endpoint.covers_addr(dst_ip))
+            {
+                return Some(peer_pif);
+            }
+        }
+        None
+    }
+}
+
+#[tracing::instrument(level = "trace")]
+fn get_src_addr(net: &Net) -> IpAddr {
+    match net {
+        Net::Ipv4(hdr) => IpAddr::V4(hdr.source().inner()),
+        Net::Ipv6(hdr) => IpAddr::V6(hdr.source().inner()),
+    }
+}
+
+#[tracing::instrument(level = "trace")]
+fn get_dst_addr(net: &Net) -> IpAddr {
+    match net {
+        Net::Ipv4(hdr) => IpAddr::V4(hdr.destination()),
+        Net::Ipv6(hdr) => IpAddr::V6(hdr.destination()),
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum NatDirection {
+    #[allow(dead_code)]
+    SrcNat,
+    #[allow(dead_code)]
+    DstNat,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum NatMode {
+    Stateless,
+    #[allow(dead_code)]
+    Stateful,
+}
+
+#[derive(Debug)]
+pub struct Nat {
+    context: GlobalContext,
```

Comment:
I believe that the global context should not be owned by the NAT module but just be referred to it somehow? The VPCs, pifs, policies, etc.. pertain to a wider scope in general IMO.
Also, if we plan to support multiple NAT stages (since your code assumes that a NAT object has a single direction; I'm assuming we'd instantiate more than one), this means that the NAT cannot own the context. Else, we'd need to replicate the context for each one?